_posts/2019-06-09-ucx-dgx.md

```diff
@@ -0,0 +1,662 @@
+    Dask how to serialize Python objects into GPU buffers.  There is code in
+    the dask/distributed repository to do this for Numba, CuPy, and RAPIDS cuDF
+    objects, but we've really only tested CuPy seriously.  We should expand
+    this by some of the following steps:
+
+    1.  Try a distributed Dask cuDF join computation
+
+        See [dask/distributed #2746](https://github.com/dask/distributed/pull/2746) for initial work here.
+
+    2.  Teach Dask to serialize array GPU libraries, like PyTorch and
+        TensorFlow, or possibly anything that supports the
+        `__cuda_array_interface__` protocol.
+
+-   **Track down communication failures:** We still occasionally get
+    unexplained communication failures.  We should stress test this system to
+    discover rough corners.
+
+-  **TCP**: Groups with high performing TCP networks can't yet make use of UCX+Dask (though they can use either one individually).
+
+    Currently using UCX in a client-server mode as we're doing with
+    Dask requires access to RDMA libraries, which are often not found on systems
+    without networking systems like InfiniBand.  This means that groups with
+    high performing TCP networks can't make use of UCX+Dask.
+
+    This is in progress at [openucx/ucx #3570](https://github.com/openucx/ucx/pull/3570)
+
+-  **Commodity Hardware**: Currently this code is only really useful on
+   high performance Linux systems that have InfiniBand or NVLink.  However,
+   it would be nice to also use this on more commodity systems, including
+   personal laptop computers using TCP and shared memory.
+
+   Currently Dask uses TCP for inter-process communication on a single machine.
+   Using UCX on a personal computer would give us access to shared memory
+   speeds, which tend to be an order of magnitude faster.
+
+   See [openucx/ucx #3663](https://github.com/openucx/ucx/issues/3663) for more
+   information.
+
+-   **Tune Performance:**  The 5-10 GB/s bandwidths that we see with NVLink
+    today are sub-optimal.  With UCX-Py alone we're able to get something like
+    15 GB/s on large message tranfers.  We should benchmark and tune our
+    implementation to see what is taking up the extra time.  Until things work
+    more robustly though, this is a secondary priority.
+
+
+Appendix: Setup
+===============
+
+Performing these experiments depends currently on development branches in a few
+repositories.  This section includes my current setup.
+
+Create Conda Environment
+------------------------
+
+```
+conda create -n ucx python=3.7 libtool cmake automake autoconf cython bokeh pytest pkgconfig ipython dask numba -y
+```
+
+Note: for some reason using conda-forge makes the autogen step below fail.
+
+Set up UCX
+----------
+
+```
+# Clone UCX repository and get branch
+git clone https://github.com/openucx/ucx
+cd ucx
+git remote add Akshay-Venkatesh git@github.com:Akshay-Venkatesh/ucx.git
```

Comment:
Do you want to change these as well ?

Reply:
Nope.  That's Akshay's fork of the `ucx` repository, not ucx-py

Reply:
ah right, sorry for the noise